File: examples/renderer-demo/crates/core/src/events/events.rs
use num_derive::FromPrimitive;    
use num_traits::FromPrimitive;
use serde::{Serialize, Deserialize};
use std::convert::TryFrom;

//the order must match typescript!
#[derive(FromPrimitive)]
#[repr(u32)]
pub enum BridgeEventIndex {
    WindowSize,
    LoadGltf,
    GltfLoaded,
    CameraSettings,
}

//Let's us get a BridgeEvent from the number which is sent from JS
impl TryFrom<u32> for BridgeEventIndex {
    type Error = &'static str;

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        FromPrimitive::from_u32(value).ok_or("BridgeEvent: outside of range!")
    }
}

//All the event data:

#[derive(Serialize, Deserialize)]
pub struct WindowSize {
    pub width: u32,
    pub height: u32 
}


#[derive(Serialize, Deserialize)]
pub struct CameraSettings {
    pub style: u32,
    pub xmag: Option<f64>,
    pub ymag: Option<f64>,
    pub znear: Option<f64>,
    pub zfar: Option<f64>,
    pub aspectRatio: Option<f64>,
    pub yfov: Option<f64>,
}


#[derive(FromPrimitive)]
#[repr(u32)]
pub enum CameraStyle {
    Orthographic,
    Perspective
}
impl TryFrom<u32> for CameraStyle {
    type Error = &'static str;

    fn try_from(value: u32) -> Result<Self, Self::Error> {
        FromPrimitive::from_u32(value).ok_or("CameraStyle: outside of range!")
    }
}

File: examples/renderer-demo/typescript/events/events.ts
import {set_state} from "@state/state";
/**
 * Customize this for all the bridge event types
 * If there are any complex objects, create structs on the Rust side too!
 */

//The order of these must match the Rust BridgeEvent!
export enum BridgeEvent {
    WindowSize,
    LoadGltf,
    GltfLoaded,
    CameraSettings,
    Clear
}

type ValidEvent = 
    | [BridgeEvent.WindowSize, WindowSize]
    | BridgeEvent.GltfLoaded
    | [BridgeEvent.LoadGltf, string]
    | [BridgeEvent.CameraSettings, OrthographicCameraSettings | PerspectiveCameraSettings]
    | BridgeEvent.Clear

interface WindowSize{
    width: number;
    height: number;
}

export interface OrthographicCameraSettings {
    style: CameraStyle,
    xmag: number,
    ymag: number,
    znear: number,
    zfar: number,
    positionX: number,
    positionY: number,
    positionZ: number,
}

export interface PerspectiveCameraSettings {
    style: CameraStyle,
    aspectRatio: number;
    yfov: number;
    znear: number,
    zfar: number,
    positionX: number,
    positionY: number,
    positionZ: number,
}
export enum CameraStyle {
    ORTHOGRAPHIC,
    PERSPECTIVE
}


//this is loosely defined because the types are converted on the rust side 
type EventSender = (evt_type:number, evt_data:any) => unknown;
let send_event_to_rust:EventSender;


export const send_bridge_event_from_rust_to_ts_unchecked = (evt_type:BridgeEvent, evt_data?:any) => {
    switch(evt_type) {
        case BridgeEvent.GltfLoaded: set_state("loaded"); break;
    }
}
export const send_event = (event:ValidEvent) => {
    send_event_to_rust(event[0], event[1]);
}

export const register_event_sender = (_send_event_to_rust:EventSender) => {
    send_event_to_rust = _send_event_to_rust;
}